### PR TITLE
[WIP][CR] Mutation Triggers - Overhaul and Add Clothing Based Triggers

### LINKED ISSUE #43378: Compound eyes perception bonus does not match up to its description
# Describe the bug

The compound eyes mutation's description says that it will give 2 bonus perception, but the actual stat boost is only 1.

# Steps To Reproduce

1. Acquire the mutation
2. Check the bonus

# Expected behavior

Either change the description to the real value, or change the value to match the description. I have no idea which of the two was the intended bonus, so I'll leave that to the experts.

# Screenshots

image

You can see that the bonus is giving only 1 perception, and I have no gear that covers the eyes

# Versions and configuration

- OS: Windows
    - OS Version: 10.0 1809
- Game Version: 0.E-5518-g9dbd68a [64-bit]
- Graphics Version: Tiles
- Game Language: English [en]
- Mods loaded: [
    Dark Days Ahead [dda],
    Disable NPC Needs [no_npc_food],
    Alternative Map Key [alt_map_key],
    Mythical Martial Arts [MMA],
    C.R.I.T Expansion Mod [crt_expansion],
    Crazy Cataclysm [crazy_cataclysm]
]

### PR DESCRIPTION
**Pull request by Theundyingcode**
#### Summary
SUMMARY: Infrastructure "Mutation triggers rework + add clothing based triggers."

#### Purpose of change
Successor of #43442 
fixes #43378

Not noticing the hard coded stuff in avatar.cpp, I tried fixing compound_eyes by adding a new trigger only to end up overhauling the whole mutation trigger system.

New type of clothing based mutation trigger greatly adds to the possibilities of the current mutation system and, allows for JSONizing some hard coded mutations.  Additionally, fixing up the existing reflex triggers should result in a performance boost for heavily mutated characters.

#### Describe the solution

**Basic Overview:**
- Split mutation triggers into two types; the existing "reflex" triggers and new "clothing" triggers without breaking existing mutation JSONs.
- Stop Character::process_turn from blindly checking all their mutations for triggers by caching a character's mutations with reflex triggers in my_reflex_mutations
- Add clothing trigger check to Character::calc_encumbrance using list of character's mutations w/ clothing triggers cached in my_clothing_mutations.
- Make clothing trigger checks only consider clothing encumbrance to prevent creating paradox loops due to mutations that add encumbrance.

**Clothing Mutation Trigger**
Uses same basic structure as existing trigger but takes additional array "bodyparts."  Sums the results for each body part before comparing to the thresholds.  I added a couple basic trigger types for now but, there are a few more I plan to add in a later PR (coverage, conductivity, ridged, ect.)
- ENCUMBRANCE -  Sum of encumbrance values for given body parts.
- WEARING - Number of given body parts wearing clothing / items
- WARMTH - Sum of warmth values for given parts

**TODO:**
- [ ] Update documentation: JSON_INFO.md
- [ ] Incorporate review suggestions / fixes for the inevitable multitude of overlooked issues.
- [x] Replace current trait_id -> trait -> trigger lookup in Character::check_mutation_trigger with something better... maybe make my_XXX_mutations a map?
- [ ] Figure out how cdda tests work and, fix json_test.cpp + any other tests which I broke.
- [ ] Write some additional tests.
- [ ] Figure out why recalc_speed is throwing division by 0 exceptions since I didn't mess with that (at least not directly)
stacktrace

#### Describe alternatives you've considered

To make the ENCUMBRANCE clothing trigger only use item encumbrance, I made the trigger check optionally take a data parameter but, I'm not convinced it's actually better the changing Character::calc_encumbrance to:
``` 
std::map<bodypart_id, encumbrance_data> enc;
item_encumb( enc, new_item );

for( const std::pair<const bodypart_id, encumbrance_data> &elem : enc ) {
        set_part_encumbrance_data( elem.first, elem.second );
check_mutation_clothing_triggers( );
mut_cbm_encumb( enc );
for( const std::pair<const bodypart_id, encumbrance_data> &elem : enc ) {
       set_part_encumbrance_data( elem.first, elem.second );
}
```
and then using `guy.encumb( bp )` wouldn't be better

#### Testing
My initial goal is to make a completely JSONized COMPOUND_EYES work properly and, will expand the tests afterwards.

#### Additional context

I tried naming things as reasonably as I could but have no attachment any of it if someone has a better idea.  After all the 2 things all CS people struggle with are naming things, cache durations and, off by 1 errors.

Having now actually looked at the mutations system it seems to have an excessive number of hardcoded mutation checks spewn about all over the place.  This maybe my next project.

### CODE REVIEW

**Review by anothersimulacrum — COMMENTED** [review]
I did not very thoroughly review this, so I fully expect I missed something, but I don't really see why there is templating happening here, it doesn't really seem to be used much at all, and where it is it doesn't seem to need templating.

> Having now actually looked at the mutations system it seems to have an excessive number of hardcoded mutation checks spewn about all over the place. This my next project.

Awesome, it'll be great to clean this up! If you're not already aware, trait flags are a partial solution to this issue.

**Comment by Theundyingcode** [comment]
> I did not very thoroughly review this, so I fully expect I missed something, but I don't really see why there is templating happening here, it doesn't really seem to be used much at all, and where it is it doesn't seem to need templating.

Template is mostly so I can have one check_mutation_trigger function that automatically selects which is_trigger_true function based on type.  I could get some more use out of it by eliminating the check_mutation_XXX_triggers and just calling check_mutation_trigger directly but, i'm gonna leave it as is at least until i'm further along.

**Comment by anothersimulacrum** [comment]
> Template is mostly so I can have one check_mutation_trigger function that automatically selects which is_trigger_true function based on type. I could get some more use out of it by eliminating the check_mutation_XXX_triggers and just calling check_mutation_trigger directly but, i'm gonna leave it as is at least until i'm further along.

I don't see why you need templates for this, when you can just use virtual functions.

File: src/mutation.cpp [hunk 1]
```diff
@@ -259,38 +286,79 @@ void Character::mutation_reflex_trigger( const trait_id &mut )
...
+template<typename U>
+bool clothing_trigger_data::is_trigger_true( const Character &guy, const U &data ) const
+{
+    //static_assert(typeid(std::map<bodypart_id, encumbrance_data>) == typeid(U), "Unexpected data type received.");
+    bool activate = false;
+
+    int var = 0;
+    switch( trigger ) {
+        case clothing_trigger_type::ENCUMBRANCE:
+            for( bodypart_id bp : bodyparts ) {
```
Comment (Fris0uman) [inline src/mutation.cpp:345]:
```suggestion
            for( const bodypart_id &bp : bodyparts ) {
```

File: src/mutation.cpp [hunk 2]
```diff
@@ -259,38 +286,79 @@ void Character::mutation_reflex_trigger( const trait_id &mut )
...
+                var += data.at( bp ).encumbrance;
+            }
+            break;
+        case clothing_trigger_type::WEARING:
+            for( bodypart_id bp : bodyparts ) {
```
Comment (Fris0uman) [inline src/mutation.cpp:350]:
```suggestion
            for( const bodypart_id &bp : bodyparts ) {
```

File: src/mutation.cpp [hunk 3]
```diff
@@ -259,38 +286,79 @@ void Character::mutation_reflex_trigger( const trait_id &mut )
...
+                var += guy.wearing_something_on( bp ) ? 1 : 0;
+            }
+            break;
+        case clothing_trigger_type::WARMTH:
+            for( bodypart_id bp : bodyparts ) {
```
Comment (Fris0uman) [inline src/mutation.cpp:355]:
```suggestion
            for( const bodypart_id &bp : bodyparts ) {
```

File: src/mutation.cpp [hunk 4]
```diff
@@ -305,6 +373,7 @@ bool reflex_activation_data::is_trigger_true( const Character &guy ) const
     return activate;
 }
 
+
```
Comment (Fris0uman) [inline src/mutation.cpp:376]:
```suggestion
```

**Comment by Theundyingcode** [comment]
>I don't see why you need templates for this, when you can just use virtual functions.
Took a look at the conversion to using virtual functions.  Short story is they aren't compatible with the arbitrary data param.  So, I would need to:
- Define the type which would be "anti-future proofing" (is that a term?) and thus counterproductive especially since I'm basically laying down the foundation to add triggers types later that also might want to pass data (eg. triggers on_hit)
- Do away with the data pram entirely with the solution mentioned in "describe alternatives."  The result would be a less robust, somewhat inefficient but significantly simpler implementation.
- Learn something new about C++ programing that would allow a pure virtual function to take a parameter of arbitrary type.

**Comment by anothersimulacrum** [comment]
Okay I see, I was missing that you were using it to pass an arbitrary type, my bad.

**Comment by Theundyingcode** [comment]
So, i've traced my division by zero errors to: 
`Character::weight_capacity()` returning 0
Because in
`calc_mutation_value_multiplicative<&mutation_branch::weight_capacity_modifier>( cached_mutations )`

The values for mut->*member appear to be cata:optionals w/ full=true and value=(what looks like a nullptr).  When full should be false.

Still haven't figured out why that's happening though.

**Pull request closed without merging**